url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/XMLDecoder.java

```diff
@@ -0,0 +1,705 @@
+                    elem.result = idObjMap.get(elem.idref);
+                } else if (elem.isBasicType) {
+                    elem.result = executeBasic(elem);
+                } else {
+                    elem.result = executeCommon(elem);
+                }
+            } catch (Exception e) {
+                listener.exceptionThrown(e);
+            }
+
+            // track id
+            if (elem.id != null) {
+                idObjMap.put(elem.id, elem.result);
+            }
+
+            elem.isExecuted = true;
+            return elem.result;
+        }
+
+        @SuppressWarnings("nls")
+        private Object executeCommon(Elem elem) throws Exception {
+            // pop args
+            ArrayList<Object> args = new ArrayList<Object>(5);
+            while (readObjs.peek() != elem) {
+                Elem argElem = readObjs.pop();
+                args.add(0, argElem.result);
+            }
+            // decide method name
+            String method = elem.methodName;
+            if (elem.fromProperty) {
+                method = (args.size() == 0 ? "get" : "set")
+                        + capitalize(method);
+            }
+            if (elem.fromIndex) {
+                Integer index = Integer.valueOf(method);
+                args.add(0, index);
+                method = args.size() == 1 ? "get" : "set";
+            }
+            if (elem.fromField) {
+                Field f = ((Class<?>) elem.target).getField(method);
+                return (new Expression(f, "get", new Object[] { null }))
+                        .getValue();
+            }
+            if (elem.fromOwner) {
+                return owner;
+            }
+
+            if (elem.target == owner) {
+                if ("getOwner".equals(method)) {
+                    return owner;
+                }
+                Class<?>[] c = new Class[args.size()];
+                for (int i = 0; i < args.size(); i++) {
+                    Object arg = args.get(i);
+                    c[i] = (arg == null ? null: arg.getClass());
+                }
+
+                // Try actual match method
+                try {
+                    Method m = owner.getClass().getMethod(method, c);
+                    return m.invoke(owner, args.toArray());
+                } catch (NoSuchMethodException e) {
+                    // Do nothing
+                }
+
+                // Find the specific method matching the parameter
+                Method mostSpecificMethod = findMethod(
+                        owner instanceof Class<?> ? (Class<?>) owner : owner
+                                .getClass(), method, c);
+
+                return mostSpecificMethod.invoke(owner, args.toArray());
+            }
+
+            // execute
+            Expression exp = new Expression(elem.target, method, args.toArray());
+            return exp.getValue();
+        }
+
+        private Method findMethod(Class<?> clazz, String methodName,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 398-419, lines 433-452 into separate methods.